File: apps/tauri/KawaCAD/src/features/workspace/selectors/appErrorPresentation.ts
import { appStrings } from "@/localization";

export type AppErrorCategory = "userCorrectable" | "operationFailure" | "systemInternal";

export type AppErrorIdentity = {
  category: AppErrorCategory;
  code: string;
  operation: string;
  commandKind?: string;
  constraintKind?: string;
  targetIds: string[];
};

export type AppErrorPresentation = {
  id: string;
  identity: AppErrorIdentity;
  message: string;
  details?: string;
  recoverySuggestion?: string;
  occurrenceCount: number;
};

type ErrorContext = {
  category?: AppErrorCategory;
  code?: string;
  operation: string;
  commandKind?: string;
  constraintKind?: string;
  targetIds?: string[];
  recoverySuggestion?: string;
};

function recordValue(value: unknown): Record<string, unknown> | undefined {
  return typeof value === "object" && value !== null ? (value as Record<string, unknown>) : undefined;
}

function errorRecord(error: unknown) {
  const record = recordValue(error);
  if (record) return record;
  if (typeof error !== "string") return undefined;
  try {
    return recordValue(JSON.parse(error));
  } catch {
    return undefined;
  }
}

function errorText(error: unknown) {
  if (error instanceof Error) return String(error);
  const record = errorRecord(error);
  if (typeof error === "string" && !record) return error;
  return typeof record?.message === "string" ? record.message : String(error);
}

export function makeAppErrorPresentation(error: unknown, context: ErrorContext): AppErrorPresentation {
  const errorData = errorRecord(error);
  const detailsRecord = recordValue(errorData?.details);
  const category = context.category ?? "operationFailure";
  const code = context.code ?? (typeof errorData?.code === "string" ? errorData.code : category);
  const commandKind =
    context.commandKind ?? (typeof detailsRecord?.commandKind === "string" ? detailsRecord.commandKind : undefined);
  const constraintKind =
    context.constraintKind ??
    (typeof detailsRecord?.constraintKind === "string" ? detailsRecord.constraintKind : undefined);
  const targetIds = (
    context.targetIds ??
    (Array.isArray(detailsRecord?.targetIds)
      ? detailsRecord.targetIds.filter((value): value is string => typeof value === "string")
      : [])
  ).sort();
  const identity: AppErrorIdentity = {
    category,
    code,
    operation: context.operation,
    commandKind,
    constraintKind,
    targetIds,
  };
  const id = [
    identity.category,
    identity.code,
    identity.operation,
    identity.commandKind ?? "",
    identity.constraintKind ?? "",
    identity.targetIds.join(","),
  ].join("|");

  return {
    id,
    identity,
    message: errorText(error),
    details:
      typeof errorData?.details === "string"
        ? errorData.details
        : errorData?.details
          ? JSON.stringify(errorData.details)
          : undefined,
    recoverySuggestion: context.recoverySuggestion,
    occurrenceCount: 1,
  };
}

export function mergeAppErrorPresentation(
  current: AppErrorPresentation | undefined,
  next: AppErrorPresentation,
): AppErrorPresentation {
  return current?.id === next.id ? { ...next, occurrenceCount: current.occurrenceCount + 1 } : next;
}

export function appErrorCategoryTitle(category: AppErrorCategory) {
  return appStrings.error.category[category];
}

File: apps/macos/KawaCAD/Sources/KawaCADApp/Features/Workspace/Selectors/AppErrorPresentation.swift
import Foundation

enum AppErrorCategory: String, Hashable {
  case userCorrectable
  case operationFailure
  case systemInternal

  var title: String {
    switch self {
    case .userCorrectable:
      return AppStrings.tr("error.category.user_correctable")
    case .operationFailure:
      return AppStrings.tr("error.category.operation_failure")
    case .systemInternal:
      return AppStrings.tr("error.category.system_internal")
    }
  }

}

struct AppErrorIdentity: Hashable {
  let category: AppErrorCategory
  let code: String
  let operation: String
  let commandKind: String?
  let constraintKind: String?
  let targetIDs: [String]
}

struct AppErrorPresentation: Identifiable, Hashable {
  let identity: AppErrorIdentity
  let message: String
  let details: String?
  let recoverySuggestion: String?
  var occurrenceCount: Int

  var id: String {
    [
      identity.category.rawValue,
      identity.code,
      identity.operation,
      identity.commandKind ?? "",
      identity.constraintKind ?? "",
      identity.targetIDs.joined(separator: ","),
    ].joined(separator: "|")
  }
}

extension AppErrorPresentation {
  static func coreFailure(
    _ failure: CoreFailure,
    operation: String,
    commandKind: String? = nil
  ) -> AppErrorPresentation {
    let details = failure.details?.objectValue ?? [:]
    return make(
      category: .operationFailure,
      code: failure.code,
      operation: operation,
      message: failure.localizedDescription,
      details: failure.message,
      commandKind: commandKind ?? details["commandKind"]?.stringValue,
      constraintKind: details["constraintKind"]?.stringValue,
      targetIDs: details["targetIds"]?.stringArrayValue ?? []
    )
  }

  static func make(
    category: AppErrorCategory,
    code: String,
    operation: String,
    message: String,
    details: String? = nil,
    recoverySuggestion: String? = nil,
    commandKind: String? = nil,
    constraintKind: String? = nil,
    targetIDs: [String] = []
  ) -> AppErrorPresentation {
    AppErrorPresentation(
      identity: AppErrorIdentity(
        category: category,
        code: code,
        operation: operation,
        commandKind: commandKind,
        constraintKind: constraintKind,
        targetIDs: targetIDs.sorted()
      ),
      message: message,
      details: details,
      recoverySuggestion: recoverySuggestion,
      occurrenceCount: 1
    )
  }
}
